src/mongoc/mongoc-topology-description.c

```diff
@@ -0,0 +1,996 @@
+ *  }
+ *
+ *--------------------------------------------------------------------------
+ */
+
+typedef void (*transition_t)(mongoc_topology_description_t *topology,
+                             mongoc_server_description_t   *server);
+
+transition_t
+gSDAMTransitionTable[MONGOC_SERVER_DESCRIPTION_TYPES][MONGOC_TOPOLOGY_DESCRIPTION_TYPES] = {
+   { /* UNKNOWN */
+      NULL, /* MONGOC_TOPOLOGY_UNKNOWN */
+      NULL, /* MONGOC_TOPOLOGY_SHARDED */
+      NULL, /* MONGOC_TOPOLOGY_RS_NO_PRIMARY */
+      _mongoc_topology_description_check_if_has_primary /* MONGOC_TOPOLOGY_RS_WITH_PRIMARY */
+   },
+   { /* STANDALONE */
+      _mongoc_topology_description_update_unknown_with_standalone,
+      _mongoc_topology_description_remove_server,
+      _mongoc_topology_description_remove_server,
+      _mongoc_topology_description_remove_and_check_primary
+   },
+   { /* MONGOS */
+      _mongoc_topology_description_set_topology_type_to_sharded,
+      NULL,
+      _mongoc_topology_description_remove_server,
+      _mongoc_topology_description_remove_and_check_primary
+   },
+   { /* PRIMARY */
+      _mongoc_topology_description_update_rs_from_primary,
+      _mongoc_topology_description_remove_server,
+      _mongoc_topology_description_update_rs_from_primary,
+      _mongoc_topology_description_update_rs_from_primary
+   },
+   { /* SECONDARY */
+      _mongoc_topology_description_transition_unknown_to_rs_no_primary,
+      _mongoc_topology_description_remove_server,
+      _mongoc_topology_description_update_rs_without_primary,
+      _mongoc_topology_description_update_rs_with_primary_from_member
+   },
+   { /* ARBITER */
+      _mongoc_topology_description_transition_unknown_to_rs_no_primary,
+      _mongoc_topology_description_remove_server,
+      _mongoc_topology_description_update_rs_without_primary,
+      _mongoc_topology_description_update_rs_with_primary_from_member
+   },
+   { /* RS_OTHER */
+      _mongoc_topology_description_transition_unknown_to_rs_no_primary,
+      _mongoc_topology_description_remove_server,
+      _mongoc_topology_description_update_rs_without_primary,
+      _mongoc_topology_description_update_rs_with_primary_from_member
+   },
+   { /* RS_GHOST */
+      NULL,
+      _mongoc_topology_description_remove_server,
+      NULL,
+      _mongoc_topology_description_check_if_has_primary
+   }
+};
+
+/*
+ *--------------------------------------------------------------------------
+ *
+ * Handle an ismaster. This is called by the background SDAM process.
+ *
+ *--------------------------------------------------------------------------
+ */
+
+bool
+mongoc_topology_description_handle_ismaster (
+   mongoc_topology_description_t *topology,
+   mongoc_server_description_t   *sd,
+   const bson_t                  *ismaster_response,
+   int64_t                        rtt_msec,
+   bson_error_t                  *error)
+{
+   bson_return_val_if_fail (topology, false);
+   bson_return_val_if_fail (sd, false);
+
```

Comment:
i think here you can return if `(!_mongoc_topology_description_topology_has_server(...)) { return false; }` here -- we should add a server to the topology description *before* we start monitoring it, so if it's not in the topology description that means the primary told us to stop monitoring it while we awaited this response, so we can ignore this response. this is a refactoring: you already check `_mongoc_topology_description_topology_has_server` where it matters elsewhere, but you can pull it up to this level. See:

https://github.com/mongodb/mongo-python-driver/blob/3.0-dev/pymongo/topology.py#L131

PyMongo doesn't even parse an ismaster response from a server if another thread has removed the server from the topology description while we awaited its response.